include/aspect/utilities.h

```diff
@@ -1011,6 +1012,140 @@ namespace aspect
+     * The order of the following data columns has to be
+     * 'coordinates data' with @p querydim coordinate columns and @p components
+     * data columns. @p querydim and @p components need to
+     * match NUMBERCOORDCOLUMNS NUMBERDATACOLUMNS, respectivily.
+     * The coordinates do not need to be equidistant.
+     */
+    template <int querydim>
+    class UnstructuredDataLookup
+    {
+      public:
+
+        /**
+         * The following are member variables and functions of the class
+         */
+        unsigned int n_data_files;
+
+        std::vector< Point<querydim> > coordinate_values;
+
+        std::vector< std::vector<double> > data;
+
+
+        // Create an instance of a kdtree object as a member variable of the class
+        KDTree<querydim> coordinate_kdtree;
+
+
+        virtual void initialize (const MPI_Comm &comm,
+                                 std::string datadirectory,
+                                 std::vector<std::string> file_name_list);
+
+
+
+        /**
+         * Constructor that explicitly prescribes the number of data columns
+         * in the data file. If a list of data components is provided in the
+         * data file it is checked that the length of this list is consistent
+         * with this number of components. This constructor is mostly provided
+         * for backwards compatibility. Not prescribing the number of components
+         * and instead reading them from the input file allows for more
+         * flexible files.
+         */
+        UnstructuredDataLookup(const unsigned int components);
+
+        /**
+         * This constructor relies on the list of column names at the beginning
+         * of the model file to determine the number of data components,
+         * therefore when using this constructor it is necessary to provide
+         * this list in the first uncommented line of the data file.
+         */
+        UnstructuredDataLookup();
+
+        /**
+         * Loads a data text file. Throws an exception if the file does not
+         * exist, if the data file format is incorrect or if the file grid
+         * changes over model runtime.
+         */
+        void
+        load_file(const std::string &filename,
+                  const MPI_Comm &communicator);
+
+
+        /**
+         * Returns the data (e.g., velocity, temperature, etc. - according
+         * to the used plugin) for the n closest points to the target point.
+         * The function returns a vector of pairs of < data_return , Euclidean distance >,
+         * where data_return is a vector containing all the data properties
+         * associated with a single line in the file -- all the data columns.
+         * The Euclidean distance is the distance that the data points
+         * are from the target point.
+         */
+        std::vector< std::pair<std::vector<double>, double> >
+        get_data (const Point<querydim> &target_point, const unsigned int n_points) const;
```

Comment:
I think it would be great to have a second function that only returns the closest point to make the return value simpler.